### PR TITLE
Reduce database hits

### PR DESCRIPTION
**Pull request by serenasyw**

### CODE REVIEW

**Comment by SillyInventor** [comment]
Linting, poke poke.

Other than that though, it looks good. However, you will need to dive deeper to cut down the queries. I should say that you can do the list page in just three queries in about 1/7th the time.

**Comment by SillyInventor** [comment]
Also, consider looking through the changes that I never finished (due to time constraints):
https://github.com/CSDTs/django_teams/tree/reduceSQLHits

They should reduce the DB hits to O(3) on the list view...

**Comment by serenasyw** [comment]
I just found that the templates in reduceSQLHits is different from the
master branch. So which one am I supposed to use?

2018-03-01 14:07 GMT-05:00 Silly Inventor <notifications@github.com>:

**Comment by SillyInventor** [comment]
Make the changes you like. But at the moment we overwrite the
templates... so if you're going to play with them (which you'll need to
at some point), use the ones from RPI_CSDT_Community.

On 3/3/2018 9:05 PM, Yawen Shen wrote:
> I just found that the templates in reduceSQLHits is different from the
> master branch. So which one am I supposed to use?

-- 
James Davis

**Comment by SillyInventor** [comment]
Looks like it's failing some tests. You will need to change those so that it doesn't assume there is a .owners function on the teams model.

File: django_teams/models.py [hunk 1]
```diff
@@ -28,67 +29,15 @@ def __unicode__(self):
     def add_user(self, user, team_role=1):
         TeamStatus(user=user, team=self, role=team_role).save()
 
-    def owners(self):
```
Comment (SillyInventor) [inline django_teams/models.py:31]: If you remove these, you will also need to make a PR to RPI_CSDT_Community to fix the templates when you do (at some point we need to add version control).

File: django_teams/templates/django_teams/team_list.html [hunk 2]
```diff
@@ -2,10 +2,20 @@
 {% load teams %}
```
Comment (SillyInventor) [inline django_teams/templates/django_teams/team_list.html:2]: This creates duplicate entries for team owners. Which you don't want.

File: django_teams/templates/django_teams/team_detail.html [hunk 3]
```diff
@@ -12,7 +12,7 @@ <h1>{{ object }} <small>detailed information</small></h1>
             <h3 class="panel-title">Teachers</h3>
```
Comment (SillyInventor) [inline django_teams/templates/django_teams/team_detail.html:12]: NotImplementedError at /teams/4/
DISTINCT ON fields is not supported by this database backend

File: django_teams/templates/django_teams/_table.html [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+          {% if object.role == 20 %}
+              (<a href="{% url 'team-edit' pk=object.pk %}">Admin</a>)
+          {% endif %}
+      {% endif %}
+    </td>
+    <td>{% if object.description %}{{ object.description|truncatewords:15 }}{% endif %}</td>
+    <td>
+        {{ object.owner }}
+    </td>
+    <td>{{ object.member_count }}</td>
```
Comment (SillyInventor) [inline django_teams/templates/django_teams/_table.html]: In correctly counts members for classes of which you are the owner (maybe member too?).

File: django_teams/templates/django_teams/_table.html [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+      <a href="{% url 'team-detail' pk=object.id %}">{{ object }}</a>
+      {% if not user.is_anonymous %}
+          {% if object.role == 20 %}
+              (<a href="{% url 'team-edit' pk=object.pk %}">Admin</a>)
+          {% endif %}
+      {% endif %}
+    </td>
+    <td>{% if object.description %}{{ object.description|truncatewords:15 }}{% endif %}</td>
+    <td>
+        {{ object.owner }}
```
Comment (SillyInventor) [inline django_teams/templates/django_teams/_table.html]: In correctly returns owner for the bottom half of the class list.

**Review by SillyInventor — DISMISSED** [review]
:heart: it. However, there are some problems.

**Comment by serenasyw** [comment]
I just have some questions about the team list.. Is owner of a team counted
as a member too? And it seems like even if my status for a team is pending
approval, that still adds to the # of members by 1. I don't know if that is
an error.

2018-03-06 14:25 GMT-05:00 Silly Inventor <notifications@github.com>:

**Comment by SillyInventor** [comment]
The count of members should be: owners + members. Owners should not appear in the list of members.

**Comment by serenasyw** [comment]
I have figured out that the duplicate teams and incorrect count of member
result from the foreign key used in annotation. I came up with a solution
to correct the owner and # of counts but this increases the number of
queries and I'm still trying to remove the duplicate teams. I have pushed
the changes to the repo and just want to know if this is a feasible
solution...

2018-03-19 9:47 GMT-04:00 Silly Inventor <notifications@github.com>:

**Comment by SillyInventor** [comment]
Hmm, I am still getting duplicate teams in my tests. I suspect it's because the case(when()) statement for owner is creating duplicate entries.

Pretty much, as soon as you get a step in the right direction, that doesn't break anything, we can merge that. Optimization is an inch by inch process after all.

**Comment by SillyInventor** [comment]
That's some solid work. I realized that I had left some stuff undone in ownership and submission, so I fixed those for you. I think it's working. But since both of us have worked on this, I am asking @ryaholm to review it for us. Thanks in advance.

**Review by ghost — APPROVED** [review]
All features seem to work over here. LGTM. Approve. Great work @serenasyw !

**Pull request merged**